load-gen/src/main/java/io/smartcat/data/loader/datasource/CSVDataSource.java

```diff
@@ -0,0 +1,75 @@
+package io.smartcat.data.loader.datasource;
+
+import io.smartcat.data.loader.api.DataSource;
+import org.apache.commons.csv.CSVParser;
+import org.apache.commons.csv.CSVRecord;
+
+import java.util.Iterator;
+
+/**
+ * Data source that reads data from CSV file.
+ *
+ * @param <T> Type of the data provided by data source implementation.
+ */
+public class CSVDataSource<T> implements DataSource<T>, AutoCloseable {
+
+    private final CSVParser csvParser;
+    private final RowMapper<T> rowMapper;
+
+    private Iterator<CSVRecord> iterator;
+    private boolean closed = false;
+
+    /**
+     * COnstructs CSV data source with specified <code>csvParser</code> and <code>rowMapper</code>.
+     * @param csvParser parser which provides CSV records to this data source.
+     * @param rowMapper mapper which maps CSV records to type {@code <T>}.
+     */
+    public CSVDataSource(CSVParser csvParser, RowMapper<T> rowMapper) {
+        if (csvParser == null) {
+            throw new IllegalArgumentException("CSV parser cannot be null.");
+        }
+        if (rowMapper == null) {
+            throw new IllegalArgumentException("Row mapper cannot be null.");
+        }
+        this.rowMapper = rowMapper;
+        this.csvParser = csvParser;
+        this.iterator = this.csvParser.iterator();
+    }
+
+    @Override
+    public boolean hasNext(long time) {
+        if (closed) {
+            throw new IllegalStateException("Data source is already closed.");
```

Comment:
I don't agree with this. To me, #hasNext() is to be invoked on open data source. Closed data source means that no operation on it is valid. Also, this is situation is not something that can happen in 'normal use case', to achieve this, data source needs to be closed.